### ISSUE TITLE
21.11.2.2 MEMORY_LIMIT_EXCEEDED

### ISSUE THREAD

**Issue by ray-at-github** [ClickHouse/ClickHouse#31186]
After upgrading to 21.11.2.2, any complex query will result in this error:

`Code: 241. DB::Exception: Memory limit (total) exceeded: would use 20.66 GiB (attempt to allocate chunk of 4256164 bytes), maximum: -8.00 EiB. (MEMORY_LIMIT_EXCEEDED) (version 21.11.2.2 (official build))`

The strange bit is the minus maximum value.

My config (only relevant to memory settings):

`<max_server_memory_usage_to_ram_ratio>2.5</max_server_memory_usage_to_ram_ratio>`
`<max_memory_usage>600000000000</max_memory_usage>`
`
The server has 240GB RAM and a 400GB swap file.

I understand swap file shouldn't be used but my application has 1 particular query that requires about 300GB memory, which is just over my physical memory.

Immediately rolled back to 21.10.2.15 and the issue does not occur any more.

**Comment by sudhakar** [comment]
Here's the logs with v21.11.2.2 after upgrade
```
2021.11.09 15:01:56.315814 [ 1240522 ] {885ee7a1-93f4-4cae-885e-e7a193f4ccae::202111_4553_4563_1} <Debug> MergeTask::MergeProjectionsStage: Merge sorted 203 rows, containing 20 columns (20 merged, 0 gathered) in 0.026995499 sec., 7519.772092377326 rows/sec., 4.57 MiB/sec.
2021.11.09 15:01:56.316855 [ 1240510 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 827.66 MiB (attempt to allocate chunk of 4316184 bytes), maximum: -8.00 EiB. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0x9b66fd4 in /usr/bin/clickhouse
1. DB::Exception::Exception<char const*, char const*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, long&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const*&&, char const*&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, long&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&) @ 0x9b7d0c9 in /usr/bin/clickhouse
2. MemoryTracker::allocImpl(long, bool) @ 0x9b7ca43 in /usr/bin/clickhouse
3. MemoryTracker::allocImpl(long, bool) @ 0x9b7c810 in /usr/bin/clickhouse
4. MemoryTracker::allocImpl(long, bool) @ 0x9b7c810 in /usr/bin/clickhouse
5. DB::Memory<Allocator<false, false> >::alloc() @ 0x9bb75ad in /usr/bin/clickhouse
6. DB::WriteBufferFromFile::WriteBufferFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned long, int, unsigned int, char*, unsigned long) @ 0x9c6aa0c in /usr/bin/clickhouse
7. DB::DiskLocal::writeFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned long, DB::WriteMode) @ 0x11d72e97 in /usr/bin/clickhouse
8. DB::MergeTreePartition::store(DB::Block const&, std::__1::shared_ptr<DB::IDisk> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::MergeTreeDataPartChecksums&) const @ 0x12e6d1f3 in /usr/bin/clickhouse
9. DB::MergeTreePartition::store(DB::MergeTreeData const&, std::__1::shared_ptr<DB::IDisk> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::MergeTreeDataPartChecksums&) const @ 0x12e6cff1 in /usr/bin/clickhouse
10. DB::MergedBlockOutputStream::finalizePartOnDisk(std::__1::shared_ptr<DB::IMergeTreeDataPart> const&, DB::NamesAndTypesList&, DB::MergeTreeDataPartChecksums&, bool) @ 0x12ec7a3d in /usr/bin/clickhouse
11. DB::MergedBlockOutputStream::writeSuffixAndFinalizePart(std::__1::shared_ptr<DB::IMergeTreeDataPart>&, bool, DB::NamesAndTypesList const*, DB::MergeTreeDataPartChecksums*) @ 0x12ec6e5a in /usr/bin/clickhouse
12. DB::MergeTask::MergeProjectionsStage::finalizeProjectionsAndWholeMerge() const @ 0x12d46273 in /usr/bin/clickhouse
13. bool std::__1::__function::__policy_invoker<bool ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::MergeTask::MergeProjectionsStage::subtasks::'lambda1'(), bool ()> >(std::__1::__function::__policy_storage const*) @ 0x12d4a889 in /usr/bin/clickhouse
14. DB::MergeTask::MergeProjectionsStage::execute() @ 0x12d463cb in /usr/bin/clickhouse
15. DB::MergeTask::execute() @ 0x12d45f1a in /usr/bin/clickhouse
16. DB::MergePlainMergeTreeTask::executeStep() @ 0x12fc0c8c in /usr/bin/clickhouse
17. DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(std::__1::shared_ptr<DB::TaskRuntimeData>) @ 0x12d5497d in /usr/bin/clickhouse
18. DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::threadFunction() @ 0x12d5541a in /usr/bin/clickhouse
19. ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) @ 0x9babc8a in /usr/bin/clickhouse
20. ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()&&...)::'lambda'()::operator()() @ 0x9badaa7 in /usr/bin/clickhouse
21. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x9ba9097 in /usr/bin/clickhouse
22. ? @ 0x9baca9d in /usr/bin/clickhouse
23. start_thread @ 0x9609 in /usr/lib/x86_64-linux-gnu/libpthread-2.31.so
24. clone @ 0x122293 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
 (version 21.11.2.2 (official build))
```

Also CPU is getting maxed
<img width="646" alt="Screenshot 2021-11-09 at 20 49 46" src="https://user-images.githubusercontent.com/6782/140951962-89171d38-0936-4412-ada3-69a954b088ee.png">

**Comment by azat** [comment]
Seems that the problem is in #30574, it should be fixed by #31157

**Comment by ray-at-github** [comment]
Looks like it's fixed in 21.11.3.6, thanks!

**Comment by azat** [comment]
Then this can be closed.
